### PR TITLE
[flutter_tools] include LICENSE files as build dependencies

### LINKED ISSUE #27720: "flutter build apk" doesn't always update LICENSE file
Steps to reproduce:

1. `flutter create dummy; cd dummy`
2. `flutter build apk`
3. Check that `LICENSE` file does not contain "android_alarm_manager": `grep android_alarm_manager build/app/intermediates/flutter/release/flutter_assets/LICENSE`
4. Add `android_alarm_manager: any` to the pubspec.yaml file's `dependencies` section.
5. `flutter build apk`
6. Check that `build/app/intermediates/flutter/release/flutter_assets/LICENSE` now does contain "android_alarm_manager".
7. Check that `LICENSE` file does not contain "Bui": `grep Bui build/app/intermediates/flutter/release/flutter_assets/LICENSE`
8. Add `flutter_local_notifications: any` to the pubspec.yaml file's `dependencies` section.
9. `flutter build apk`
10. Check that `build/app/intermediates/flutter/release/flutter_assets/LICENSE` now does contain "Bui".
11. Run `flutter clean; flutter build apk;
12. Check that `build/app/intermediates/flutter/release/flutter_assets/LICENSE` now does contain "Bui".

For some reason, steps 10 and 12 give different results.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

have the AssetManifest track LICENSE files as input dependencies using a refactored LicenseCollector.  Surface these dependencies as depfile inputs to ensure the bundle is rebuilt if a LICENSE file changes. 

Adds note about current bug in the license processing, and includes tests for some of the current license behavior.

Fixes https://github.com/flutter/flutter/issues/27720

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Note: this needs a corresponding change in google3 due to `additionalDependencies` field added to the `AssetManifest`

File: packages/flutter_tools/lib/src/asset.dart [hunk 1]
```diff
@@ -336,78 +348,108 @@ List<_Asset> _getMaterialAssets(String fontSet) {
...
+
+    for (final String packageName in packageMap.map.keys) {
+      final Uri package = packageMap.map[packageName];
+      if (package == null || package.scheme != 'file') {
+        continue;
       }
-      if (licenseText == null) {
-        packageNames = <String>[packageName];
-        licenseText = rawLicense;
+      final File file = _fileSystem.file(package.resolve('../LICENSE'));
```
Comment (dnfield) [inline packages/flutter_tools/lib/src/asset.dart]: Not for this patch, but this seems like something we should delegate to pub.

**Review by dnfield — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
Going to avoid landing this until I update other implementations

**Comment by jonahwilliams** [comment]
Upstream change has landed, this is safe to submit now

**Pull request merged**